flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/enumerator/DynamoDbStreamsSourceEnumerator.java

```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Override
+    public void addSplitsBack(List<DynamoDbStreamsShardSplit> splits, int subtaskId) {
+        if (!splitAssignment.containsKey(subtaskId)) {
+            LOG.warn(
+                    "Unable to add splits back for subtask {} since it is not assigned any splits. Splits: {}",
+                    subtaskId,
+                    splits);
+            return;
+        }
+
+        for (DynamoDbStreamsShardSplit split : splits) {
+            splitAssignment.get(subtaskId).remove(split);
+            assignedSplitIds.remove(split.splitId());
+            unassignedSplits.add(split);
+        }
+
+        // Assign the unassignedSplits
+        // We did not discover any new splits, so we put in an empty list
+        assignSplits(Collections.emptyList(), null);
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        splitAssignment.putIfAbsent(subtaskId, new HashSet<>());
+    }
+
+    @Override
+    public DynamoDbStreamsSourceEnumeratorState snapshotState(long checkpointId) throws Exception {
+        return new DynamoDbStreamsSourceEnumeratorState(unassignedSplits, lastSeenShardId);
+    }
+
+    @Override
+    public void close() throws IOException {
+        streamProxy.close();
+    }
+
+    private List<DynamoDbStreamsShardSplit> initialDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        return mapToSplits(shards, sourceConfig.get(STREAM_INITIAL_POSITION));
+    }
+
+    /**
+     * This method is used to discover DynamoDb Streams splits the job can subscribe to. It can be run in
+     * parallel, is important to not mutate any shared state.
+     *
+     * @return list of discovered splits
+     */
+    private List<DynamoDbStreamsShardSplit> periodicallyDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+
+        // Any shard discovered after the initial startup should be read from the start, since they
+        // come from resharding
+        return mapToSplits(shards, InitialPosition.TRIM_HORIZON);
+    }
+
+    private List<DynamoDbStreamsShardSplit> mapToSplits(
+            List<Shard> shards, InitialPosition initialPosition) {
+        StartingPosition startingPosition;
+        switch (initialPosition) {
+            case LATEST:
+                //TODO: Figure out how to handle this for DDB Streams
+                // For dynamodb streams If LATEST is requested, we still set the starting position to the time of
+                // startup. This way, the job starts reading from a deterministic timestamp
+                // (i.e. time of job submission), even if it enters a restart loop immediately
+                // after submission.
+                startingPosition = StartingPosition.latest();
+                break;
+            case AT_TIMESTAMP:
+                throw new UnsupportedOperationException("DDB Stremas does not support shard iterator at timestamp");
```

Comment:
typo: Streams

Reply:
How come we cannot support `AT_TIMESTAMP`?

Reply:
Straems getShardIterator only supports `TRIM_HORIZON`, `AFTER_SEQUENCE_NUMBER`, `AT_SEQUENCE_NUMBER` and `LATEST`